### ISSUE TITLE
Unable to compile qhull 2020.1 on MacOS

### ISSUE THREAD

**Issue by GiulioRomualdi** [qhull/qhull#65]
When I try to compile the library on MacOS I got the following error
```
In file included from /Users/runner/runners/2.263.0/work/bipedal-locomotion-framework/bipedal-locomotion-framework/qhull/src/libqhullcpp/QhullUser.cpp:9:
/Users/runner/runners/2.263.0/work/bipedal-locomotion-framework/bipedal-locomotion-framework/qhull/src/libqhullcpp/QhullUser.h:39:41: error: a space is required between consecutive right angle brackets (use '> >')
        std::vector< std::vector< double>> doubles_vector;  //! vectors for capturing ints and doubles
                                        ^~
                                        > >
/Users/runner/runners/2.263.0/work/bipedal-locomotion-framework/bipedal-locomotion-framework/qhull/src/libqhullcpp/QhullUser.h:40:38: error: a space is required between consecutive right angle brackets (use '> >')
        std::vector< std::vector< int>> ints_vector;
                                     ^~
                                     > >
/Users/runner/runners/2.263.0/work/bipedal-locomotion-framework/bipedal-locomotion-framework/qhull/src/libqhullcpp/QhullUser.h:80:47: error: a space is required between consecutive right angle brackets (use '> >')
        const std::vector< std::vector< double>> &doublesVector() const { return doubles_vector; }
                                              ^~
                                              > >
/Users/runner/runners/2.263.0/work/bipedal-locomotion-framework/bipedal-locomotion-framework/qhull/src/libqhullcpp/QhullUser.h:84:44: error: a space is required between consecutive right angle brackets (use '> >')
        const std::vector< std::vector< int>> &intsVector() const { return ints_vector; }
                                           ^~
                                           > >
4 errors generated.
```
This error is caused because the compiler does not use C++11 Please find [here](https://stackoverflow.com/questions/29331315/double-closing-angle-brackets-generate-syntax-error-in-specific-case) the detailed explanation.
This problem can be easily solved by adding the following patch
```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index bf7a259..b527e23 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -67,6 +67,8 @@
 project(qhull)
 cmake_minimum_required(VERSION 3.0)
 
+set(CMAKE_CXX_STANDARD 11)
+
 # Define qhull_VERSION in README.txt, Announce.txt, qh-get.htm, CMakeLists.txt
 #   qhull-zip.sh (twice), qhull-wiki.md, qhull-news.htm, File_id.diz, index.htm
 #   Makefile (SO), qhull-warn.pri (VERSION), qhull-exports.def (VERSION 8.0), qhull_p-exports.def, qhull_r-exports.def
```

Fixing this issue will also allow the porting of the new version of the library in `vcpkg` (Associated issue: https://github.com/microsoft/vcpkg/issues/12004)

Another solution is to use `target_compile_features(mylib PUBLIC cxx_std_11)` but this requires to changed the required version of `cmake` from 3.0 to 3.1.

If you are interested I can propose a PR with the first fix I proposed.

**Comment by stephanmg** [comment]
Why would you need to include C++11 support here? Just add the appropriate spaces between the consecutive `>>`, e.g. change to `> >`.

**Comment by GiulioRomualdi** [comment]
Unfortunately, this is not enough. In `user_eg3_r.cpp`  the range for loop is used and this requires `C++11`
https://github.com/qhull/qhull/blob/4733a95be0298919a646a1d6044c1bee7a552880/src/user_eg3/user_eg3_r.cpp#L95-L107

**Comment by stephanmg** [comment]
Well, this is not what you stated before. You only showed the obviously easy to fix issue in QHullUser.h.

**Comment by GiulioRomualdi** [comment]
> Well, this is not what you stated before. You only showed the obviously easy to fix issue in QHullUser.h.

I noticed while I tried to fix the problem

> I'm not sure what the maintainers opinion is, but one could replace the range based for loop with a regular for loop. Thus we do not need to enable C++11 features and are backward compatible with older compilers.

Yes I agree with you

**Comment by stephanmg** [comment]
Okay that is great, let's see what the maintainer has to say about this. I added the changes to another pull request.

**Comment by GiulioRomualdi** [comment]
Ok I'll try to recompile the code with your modification (just to understand if it solves the problem or there are other problems)

**Comment by stephanmg** [comment]
@GiulioRomualdi cool! I honestly didn't have time to test it, but I'm most certain it will work.

**Comment by GiulioRomualdi** [comment]
Unfortunately forcing `C++98` arises other errors
```
/home/gromualdi/robot-code/qhull/src/libqhullcpp/QhullUser.cpp: In member function ‘void orgQhull::QhullUser::captureOff()’:
/home/gromualdi/robot-code/qhull/src/libqhullcpp/QhullUser.cpp:101:147: error: ‘uintptr_t’ was not declared in this scope
         throw QhullError(10081, "Qhull error: conflicting QhullUser (0x%llx) for QhullUser::captureOff().  Does not match 'this' (0x...%X)", int((uintptr_t)this), 0, 0.0, qh()->cpp_user);
                                                                                                                                                   ^~~~~~~~~
/home/gromualdi/robot-code/qhull/src/libqhullcpp/QhullUser.cpp:101:147: note: suggested alternative: ‘__intptr_t’
         throw QhullError(10081, "Qhull error: conflicting QhullUser (0x%llx) for QhullUser::captureOff().  Does not match 'this' (0x...%X)", int((uintptr_t)this), 0, 0.0, qh()->cpp_user);
                                                                                                                                                   ^~~~~~~~~
                                                                                                                                                   __intptr_t
CMakeFiles/qhullcpp.dir/build.make:264: recipe for target 'CMakeFiles/qhullcpp.dir/src/libqhullcpp/QhullUser.cpp.o' failed
make[2]: *** [CMakeFiles/qhullcpp.dir/src/libqhullcpp/QhullUser.cpp.o] Error 1
CMakeFiles/Makefile2:427: recipe for target 'CMakeFiles/qhullcpp.dir/all' failed
make[1]: *** [CMakeFiles/qhullcpp.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```
`uintptr_t` has been introduced in `C++11`. See [here](https://en.cppreference.com/w/cpp/types/integer). In my opinion, the maintainers decided to use `C++11` in the release 2020.1.
Right now the code cannot compile with a compiler that supports only C++98 only (I've also tried on Ubuntu Linux `gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)`.
Let's wait for the maintainers before proceeding

**Comment by stephanmg** [comment]
I see. Well, if C++11 support is already assumed for other parts of the code base, then probably it's best to add C++11.

**Comment by GiulioRomualdi** [comment]
Associated commit https://github.com/Homebrew/homebrew-core/commit/6ca6c77df01fd41a9b0e6cca9fe0681090cd3b8a and comment https://github.com/Homebrew/homebrew-core/pull/56273#issue-434091809

**Comment by cbbarber** [comment]
Hi Giulio and Stephan,

The dependencies on C++11 were unintentional.    Can you make the various changes required to remove these dependencies?   

The C++ interface is purely for convenience.   The Qhull code is written in C and should compile with most versions.

**Comment by cbbarber** [comment]
Please let me know when you are done.   These changes will be a point release of Qhull.   C++ improvements are the main feature for Qhull 2020.1.  

Which C++ standard are you following?  I'll add it to README.txt.

Do the changes retain ABI compatibility?  If so, the release will be 8.0.1 (the current SONAME is 8.0).

**Comment by cbbarber** [comment]
The C++ files in src/qhulltest/ are dependent on Qt and C++11.  Qt has deprecated their 'foreach' macro in favor of C++11 range-based for loops.   These dependencies are OK.   qhulltest is standalone and not required for the other libraries and executables.

From Giulio's previous comment, gcc can be forced to C++98.   I'll give it a try when my computer is repaired on Tuesday (hopefully!).

**Comment by stephanmg** [comment]
Okay that sounds great. I think my changes were all C++98 compatible.

**Comment by stephanmg** [comment]
> Unfortunately forcing `C++98` arises other errors
> 
> ```
> /home/gromualdi/robot-code/qhull/src/libqhullcpp/QhullUser.cpp: In member function ‘void orgQhull::QhullUser::captureOff()’:
> /home/gromualdi/robot-code/qhull/src/libqhullcpp/QhullUser.cpp:101:147: error: ‘uintptr_t’ was not declared in this scope
>          throw QhullError(10081, "Qhull error: conflicting QhullUser (0x%llx) for QhullUser::captureOff().  Does not match 'this' (0x...%X)", int((uintptr_t)this), 0, 0.0, qh()->cpp_user);
>                                                                                                                                                    ^~~~~~~~~
> /home/gromualdi/robot-code/qhull/src/libqhullcpp/QhullUser.cpp:101:147: note: suggested alternative: ‘__intptr_t’
>          throw QhullError(10081, "Qhull error: conflicting QhullUser (0x%llx) for QhullUser::captureOff().  Does not match 'this' (0x...%X)", int((uintptr_t)this), 0, 0.0, qh()->cpp_user);
>                                                                                                                                                    ^~~~~~~~~
>                                                                                                                                                    __intptr_t
> CMakeFiles/qhullcpp.dir/build.make:264: recipe for target 'CMakeFiles/qhullcpp.dir/src/libqhullcpp/QhullUser.cpp.o' failed
> make[2]: *** [CMakeFiles/qhullcpp.dir/src/libqhullcpp/QhullUser.cpp.o] Error 1
> CMakeFiles/Makefile2:427: recipe for target 'CMakeFiles/qhullcpp.dir/all' failed
> make[1]: *** [CMakeFiles/qhullcpp.dir/all] Error 2
> Makefile:160: recipe for target 'all' failed
> make: *** [all] Error 2
> ```
> 
> `uintptr_t` has been introduced in `C++11`. See [here](https://en.cppreference.com/w/cpp/types/integer)

So, are these all instances of further sources of error? If so, since the `uintptr_t` is casted to int anyway, can't we use an `int*`?

**Comment by cbbarber** [comment]
Fixed in Qhull 8.0.1 as summarized in the Wiki

The Makefile invokes gcc with '-std=c++98'
